File: TaskManagerCLI/Menu.cpp
#include "Menu.h"
#include "Utils.h"
#include <iostream>

Menu::Menu(TaskManager& manager) : taskManager(manager) {}

void Menu::showMainMenu() const {
    std::cout << "\n=== Task Manager ===\n";
    std::cout << "1. Add Task\n";
    std::cout << "2. View All Tasks\n";
    std::cout << "3. Update Task\n";
    std::cout << "4. Delete Task\n";
    std::cout << "5. Mark Task as Complete\n";
    std::cout << "6. Exit\n";
    std::cout << "Choose an option: ";
}

void Menu::handleAddTask() {
    std::string title, description, dueDate;
    int priorityInput;

    std::cout << "Enter title: ";
    std::getline(std::cin, title);

    std::cout << "Enter description: ";
    std::getline(std::cin, description);

    std::cout << "Enter due date (YYYY-MM-DD): ";
    std::getline(std::cin, dueDate);

    if (!Utils::isValidDate(dueDate)) {
        std::cout << "Invalid date format!\n";
        return;
    }

    std::cout << "Enter priority (0: Low, 1: Medium, 2: High): ";
    std::cin >> priorityInput;
    std::cin.ignore(); // to flush newline

    if (priorityInput < 0 || priorityInput > 2) {
        std::cout << "Invalid priority!\n";
        return;
    }

    taskManager.addTask(title, description, dueDate, static_cast<TaskPriority>(priorityInput));
    std::cout << "Task added successfully.\n";
}

void Menu::handleViewTasks() const {
    auto tasks = taskManager.getAllTasks();
    if (tasks.empty()) {
        std::cout << "No tasks found.\n";
        return;
    }

    for (const auto& task : tasks) {
        std::cout << "---------------------\n";
        std::cout << task.toString() << "\n";
    }
}

void Menu::handleUpdateTask() {
    int id;
    std::cout << "Enter Task ID to update: ";
    std::cin >> id;
    std::cin.ignore();

    auto taskOpt = taskManager.getTaskById(id);
    if (!taskOpt.has_value()) {
        std::cout << "Task not found!\n";
        return;
    }

    std::string newTitle, newDescription, newDueDate;
    int newPriorityInput;

    std::cout << "Enter new title: ";
    std::getline(std::cin, newTitle);

    std::cout << "Enter new description: ";
    std::getline(std::cin, newDescription);

    std::cout << "Enter new due date (YYYY-MM-DD): ";
    std::getline(std::cin, newDueDate);

    if (!Utils::isValidDate(newDueDate)) {
        std::cout << "Invalid date format!\n";
        return;
    }

    std::cout << "Enter new priority (0: Low, 1: Medium, 2: High): ";
    std::cin >> newPriorityInput;
    std::cin.ignore();

    if (newPriorityInput < 0 || newPriorityInput > 2) {
        std::cout << "Invalid priority!\n";
        return;
    }

    if (taskManager.updateTask(id, newTitle, newDescription, newDueDate, static_cast<TaskPriority>(newPriorityInput))) {
        std::cout << "Task updated successfully.\n";
    } else {
        std::cout << "Failed to update task.\n";
    }
}

void Menu::handleDeleteTask() {
    int id;
    std::cout << "Enter Task ID to delete: ";
    std::cin >> id;
    std::cin.ignore();

    if (taskManager.deleteTask(id)) {
        std::cout << "Task deleted successfully.\n";
    } else {
        std::cout << "Task not found!\n";
    }
}

void Menu::handleMarkTaskComplete() {
    int id;
    std::cout << "Enter Task ID to mark as complete: ";
    std::cin >> id;
    std::cin.ignore();

    if (taskManager.markTaskComplete(id)) {
        std::cout << "Task marked as complete.\n";
    } else {
        std::cout << "Task not found!\n";
    }
}

void Menu::run() {
    int choice;
    while (true) {
        showMainMenu();
        std::cin >> choice;
        std::cin.ignore(); // flush newline character

        switch (choice) {
            case 1:
                handleAddTask();
                break;
            case 2:
                handleViewTasks();
                break;
            case 3:
                handleUpdateTask();
                break;
            case 4:
                handleDeleteTask();
                break;
            case 5:
                handleMarkTaskComplete();
                break;
            case 6:
                std::cout << "Exiting...\n";
                return;
            default:
                std::cout << "Invalid choice!\n";
        }
    }
}

File: src/main/java/com/taskmanager/Menu.java
package com.taskmanager;

import java.util.List;
import java.util.Optional;
import java.util.Scanner;

/**
 * Menu class for interacting with the TaskManager via console.
 */
public class Menu {
    private final TaskManager taskManager;
    private final Scanner scanner;

    public Menu(TaskManager manager) {
        this.taskManager = manager;
        this.scanner = new Scanner(System.in);
    }

    private void showMainMenu() {
        System.out.println("\n=== Task Manager ===");
        System.out.println("1. Add Task");
        System.out.println("2. View All Tasks");
        System.out.println("3. Update Task");
        System.out.println("4. Delete Task");
        System.out.println("5. Mark Task as Complete");
        System.out.println("6. Exit");
        System.out.print("Choose an option: ");
    }

    private void handleAddTask() {
        System.out.print("Enter title: ");
        String title = scanner.nextLine().trim();

        System.out.print("Enter description: ");
        String description = scanner.nextLine().trim();

        System.out.print("Enter due date (YYYY-MM-DD): ");
        String dueDate = scanner.nextLine().trim();

        if (!Utils.isValidDate(dueDate)) {
            System.out.println("Invalid date format!");
            return;
        }

        System.out.print("Enter priority (0: Low, 1: Medium, 2: High): ");
        int priorityInput;
        try {
            priorityInput = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Invalid priority input!");
            return;
        }

        if (priorityInput < 0 || priorityInput > 2) {
            System.out.println("Invalid priority!");
            return;
        }

        TaskPriority priority = TaskPriority.values()[priorityInput];
        taskManager.addTask(title, description, dueDate, priority);
        System.out.println("Task added successfully.");
    }

    private void handleViewTasks() {
        List<Task> tasks = taskManager.getAllTasks();
        if (tasks.isEmpty()) {
            System.out.println("No tasks found.");
            return;
        }

        for (Task task : tasks) {
            System.out.println("---------------------");
            System.out.println(task.toString());
        }
    }

    private void handleUpdateTask() {
        System.out.print("Enter Task ID to update: ");
        int id;
        try {
            id = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Invalid ID input!");
            return;
        }

        Optional<Task> taskOpt = taskManager.getTaskById(id);
        if (!taskOpt.isPresent()) {
            System.out.println("Task not found!");
            return;
        }

        System.out.print("Enter new title: ");
        String newTitle = scanner.nextLine().trim();

        System.out.print("Enter new description: ");
        String newDescription = scanner.nextLine().trim();

        System.out.print("Enter new due date (YYYY-MM-DD): ");
        String newDueDate = scanner.nextLine().trim();

        if (!Utils.isValidDate(newDueDate)) {
            System.out.println("Invalid date format!");
            return;
        }

        System.out.print("Enter new priority (0: Low, 1: Medium, 2: High): ");
        int newPriorityInput;
        try {
            newPriorityInput = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Invalid priority input!");
            return;
        }

        if (newPriorityInput < 0 || newPriorityInput > 2) {
            System.out.println("Invalid priority!");
            return;
        }

        TaskPriority newPriority = TaskPriority.values()[newPriorityInput];
        boolean updated = taskManager.updateTask(id, newTitle, newDescription, newDueDate, newPriority);
        if (updated) {
            System.out.println("Task updated successfully.");
        } else {
            System.out.println("Failed to update task.");
        }
    }

    private void handleDeleteTask() {
        System.out.print("Enter Task ID to delete: ");
        int id;
        try {
            id = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Invalid ID input!");
            return;
        }

        boolean deleted = taskManager.deleteTask(id);
        if (deleted) {
            System.out.println("Task deleted successfully.");
        } else {
            System.out.println("Task not found!");
        }
    }

    private void handleMarkTaskComplete() {
        System.out.print("Enter Task ID to mark as complete: ");
        int id;
        try {
            id = Integer.parseInt(scanner.nextLine().trim());
        } catch (NumberFormatException e) {
            System.out.println("Invalid ID input!");
            return;
        }

        boolean marked = taskManager.markTaskComplete(id);
        if (marked) {
            System.out.println("Task marked as complete.");
        } else {
            System.out.println("Task not found!");
        }
    }

    public void run() {
        while (true) {
            showMainMenu();
            String choiceLine = scanner.nextLine().trim();
            int choice;
            try {
                choice = Integer.parseInt(choiceLine);
            } catch (NumberFormatException e) {
                System.out.println("Invalid choice!");
                continue;
            }

            switch (choice) {
                case 1:
                    handleAddTask();
                    break;
                case 2:
                    handleViewTasks();
                    break;
                case 3:
                    handleUpdateTask();
                    break;
                case 4:
                    handleDeleteTask();
                    break;
                case 5:
                    handleMarkTaskComplete();
                    break;
                case 6:
                    System.out.println("Exiting...");
                    return;
                default:
                    System.out.println("Invalid choice!");
            }
        }
    }
}